File: tools/community-tools.js
// Deterministic operational foundations for Good Neighbour profiles.
// These tools accept caller-provided records and deliberately do not claim
// that an external roster, notification service, policy repository, or civic
// source is connected until an adapter supplies one.

const asArray = (value) => Array.isArray(value) ? value : [];
const asText = (value) => typeof value === 'string' ? value.trim() : '';
const hasAll = (values, required) => required.every((item) => values.includes(item));

function foodBankTools() {
  return [{
    spec: {
      name: 'match_food_bank_shifts',
      description: 'Match provided food-bank volunteers to shifts by availability and required skills, while reporting roster gaps and inventory threshold warnings.',
      inputSchema: { json: { type: 'object', properties: {
        volunteers: { type: 'array', description: 'Records with id, name, skills, and availability.' },
        shifts: { type: 'array', description: 'Records with id, slot, capacity, and requiredSkills.' },
        inventory: { type: 'array', description: 'Records with item and quantity.' },
        thresholds: { type: 'array', description: 'Records with item and minimum quantity.' },
      }, required: ['volunteers', 'shifts'] } },
    },
    fn: async (input = {}) => {
      const volunteers = asArray(input.volunteers).map((person, index) => ({
        id: asText(person?.id) || `volunteer-${index + 1}`,
        name: asText(person?.name) || `Volunteer ${index + 1}`,
        skills: asArray(person?.skills).map(asText).filter(Boolean),
        availability: asArray(person?.availability).map(asText).filter(Boolean),
      }));
      const assigned = new Set();
      const assignments = [];
      const gaps = [];
      for (const shift of asArray(input.shifts)) {
        const shiftId = asText(shift?.id) || 'unnamed-shift';
        const slot = asText(shift?.slot);
        const requiredSkills = asArray(shift?.requiredSkills).map(asText).filter(Boolean);
        const capacity = Math.max(1, Number(shift?.capacity) || 1);
        const matches = volunteers.filter((person) => !assigned.has(person.id)
          && (!slot || person.availability.includes(slot))
          && hasAll(person.skills, requiredSkills)).slice(0, capacity);
        matches.forEach((person) => assigned.add(person.id));
        assignments.push({ shiftId, volunteers: matches.map((person) => ({ id: person.id, name: person.name })) });
        if (matches.length < capacity) gaps.push({ shiftId, missing: capacity - matches.length, requiredSkills, slot });
      }
      const inventoryGaps = asArray(input.thresholds).flatMap((threshold) => {
        const item = asText(threshold?.item);
        const minimum = Number(threshold?.minimum ?? threshold?.min);
        const current = asArray(input.inventory).find((entry) => asText(entry?.item).toLowerCase() === item.toLowerCase());
        const quantity = Number(current?.quantity) || 0;
        return item && Number.isFinite(minimum) && quantity < minimum ? [{ item, quantity, minimum }] : [];
      });
      return JSON.stringify({ assignments, rosterGaps: gaps, inventoryGaps, notifications: 'No notifications sent; connect a notification adapter after confirming assignments.' });
    },
  }];
}

function nonprofitTools() {
  return [
    {
      spec: { name: 'generate_nonprofit_template', description: 'Generate a reusable small-nonprofit template for an incident report, board memo, or grant tracker.', inputSchema: { json: { type: 'object', properties: { templateType: { type: 'string' }, organization: { type: 'string' }, details: { type: 'string' } }, required: ['templateType'] } } },
      fn: async ({ templateType, organization = '', details = '' } = {}) => {
        const type = asText(templateType).toLowerCase();
        const headings = type.includes('incident') ? ['Incident Report', 'Date and time', 'People involved', 'What happened', 'Immediate actions', 'Follow-up owner']
          : type.includes('board') ? ['Board Memo', 'Decision requested', 'Context', 'Options considered', 'Recommendation', 'Next steps']
            : ['Grant Tracker', 'Funder and grant', 'Deadline', 'Deliverables', 'Owner', 'Status and next action'];
        return [`# ${headings[0]}`, organization ? `Organization: ${organization}` : 'Organization: [add]', ...headings.slice(1).map((heading) => `## ${heading}\n${heading === 'Status and next action' ? details || '[add]' : '[add]'}`)].join('\n\n');
      },
    },
    {
      spec: { name: 'search_internal_policies', description: 'Search caller-provided bylaws or HR policy records and return only matching text with its supplied source citation.', inputSchema: { json: { type: 'object', properties: { query: { type: 'string' }, documents: { type: 'array' } }, required: ['query', 'documents'] } } },
      fn: async ({ query = '', documents = [] } = {}) => {
        const needle = asText(query).toLowerCase();
        const matches = asArray(documents).filter((doc) => `${doc?.title || ''} ${doc?.text || ''}`.toLowerCase().includes(needle));
        if (!matches.length) return 'No supplied internal policy matched. Do not infer an organizational rule; ask for the relevant document or an administrator.';
        return matches.slice(0, 5).map((doc) => `${doc.title || 'Untitled'} — ${doc.text || ''} [Source: ${doc.source || 'source not supplied'}]`).join('\n');
      },
    },
  ];
}

function mutualAidTools() {
  return [
    {
      spec: { name: 'sanitize_mutual_aid_intake', description: 'Redact common contact, precise-location, and sensitive-status details from a mutual-aid intake before sharing it.', inputSchema: { json: { type: 'object', properties: { text: { type: 'string' } }, required: ['text'] } } },
      fn: async ({ text = '' } = {}) => asText(text)
        .replace(/[\w.+-]+@[\w-]+\.[\w.-]+/gi, '[redacted contact]')
        .replace(/(?:\+?\d[\d\s().-]{7,}\d)/g, '[redacted contact]')
        .replace(/\b(?:address|location|diagnosis|medical status|immigration status|shelter)\s*:\s*[^;\n]+/gi, (match) => `${match.split(':')[0]}: [redacted]`),
    },
    {
      spec: { name: 'pair_mutual_aid_needs', description: 'Pair anonymous needs with community offers only when both records explicitly opt in to peer connection.', inputSchema: { json: { type: 'object', properties: { needs: { type: 'array' }, offers: { type: 'array' } }, required: ['needs', 'offers'] } } },
      fn: async ({ needs = [], offers = [] } = {}) => {
        const matches = [];
        for (const need of asArray(needs)) {
          if (need?.optIn !== true) continue;
          const needTags = asArray(need.tags).map((tag) => asText(tag).toLowerCase());
          const offer = asArray(offers).find((candidate) => candidate?.optIn === true && asArray(candidate.tags).some((tag) => needTags.includes(asText(tag).toLowerCase())));
          if (offer) matches.push({ needId: asText(need.id) || 'anonymous-need', offerId: asText(offer.id) || 'anonymous-offer', status: 'pending-mutual-confirmation' });
        }
        return JSON.stringify({ matches, note: 'No contact details were shared. Both participants must confirm before connection.' });
      },
    },
  ];
}

function civicTools() {
  return [{
    spec: { name: 'verify_civic_sources', description: 'Check supplied civic sources for explicit verification and official institutional URLs before grounding an answer.', inputSchema: { json: { type: 'object', properties: { sources: { type: 'array' } }, required: ['sources'] } } },
    fn: async ({ sources = [] } = {}) => {
      const checked = asArray(sources).map((source) => {
        const url = asText(source?.url);
        const officialUrl = /^https:\/\//i.test(url) && /\.(gov|gc\.ca|edu|org)(\/|$)/i.test(url);
        return { title: asText(source?.title) || 'Untitled source', url, verified: source?.verified === true && officialUrl };
      });
      return JSON.stringify({ grounded: checked.length > 0 && checked.every((source) => source.verified), sources: checked, instruction: 'If grounded is false, state the uncertainty and provide an official contact point instead of asserting the answer.' });
    },
  }];
}

export function createCommunityTools(profileId) {
  if (profileId === 'food-bank') return foodBankTools();
  if (profileId === 'nonprofit-helpdesk') return nonprofitTools();
  if (profileId === 'mutual-aid') return mutualAidTools();
  if (profileId === 'civic-knowledge') return civicTools();
  return [];
}

File: app/ForgeAgent/community_tools.py
"""Deterministic, privacy-aware operational tools for Good Neighbour roles."""

import json
import re
from collections.abc import Callable

from strands import tool


def build_community_tools(profile_id: str) -> list[Callable[..., str]]:
    if profile_id == "food-bank":
        @tool
        def match_food_bank_shifts(volunteers: list[dict], shifts: list[dict], inventory: list[dict] | None = None, thresholds: list[dict] | None = None) -> str:
            """Match supplied volunteers to shifts by availability and skills; report roster and inventory gaps."""
            assigned: set[str] = set()
            assignments: list[dict] = []
            gaps: list[dict] = []
            for index, shift in enumerate(shifts or []):
                shift_id = str(shift.get("id") or f"shift-{index + 1}")
                slot = str(shift.get("slot") or "")
                required = {str(skill).lower() for skill in shift.get("requiredSkills", [])}
                capacity = max(1, int(shift.get("capacity", 1)))
                matches = []
                for person_index, person in enumerate(volunteers or []):
                    person_id = str(person.get("id") or f"volunteer-{person_index + 1}")
                    skills = {str(skill).lower() for skill in person.get("skills", [])}
                    availability = {str(value) for value in person.get("availability", [])}
                    if person_id not in assigned and (not slot or slot in availability) and required <= skills and len(matches) < capacity:
                        assigned.add(person_id)
                        matches.append({"id": person_id, "name": person.get("name") or person_id})
                assignments.append({"shiftId": shift_id, "volunteers": matches})
                if len(matches) < capacity:
                    gaps.append({"shiftId": shift_id, "missing": capacity - len(matches), "requiredSkills": sorted(required), "slot": slot})
            inventory_gaps = []
            stock = {str(item.get("item", "")).lower(): float(item.get("quantity", 0)) for item in (inventory or [])}
            for threshold in thresholds or []:
                item = str(threshold.get("item", ""))
                minimum = float(threshold.get("minimum", threshold.get("min", 0)))
                quantity = stock.get(item.lower(), 0)
                if item and quantity < minimum:
                    inventory_gaps.append({"item": item, "quantity": quantity, "minimum": minimum})
            return json.dumps({"assignments": assignments, "rosterGaps": gaps, "inventoryGaps": inventory_gaps, "notifications": "No notifications sent; confirm assignments and connect a notification adapter."})

        return [match_food_bank_shifts]

    if profile_id == "nonprofit-helpdesk":
        @tool
        def generate_nonprofit_template(template_type: str, organization: str = "", details: str = "") -> str:
            """Generate an incident report, board memo, or grant tracker template."""
            kind = template_type.lower()
            headings = (["Incident Report", "Date and time", "People involved", "What happened", "Immediate actions", "Follow-up owner"] if "incident" in kind else ["Board Memo", "Decision requested", "Context", "Options considered", "Recommendation", "Next steps"] if "board" in kind else ["Grant Tracker", "Funder and grant", "Deadline", "Deliverables", "Owner", "Status and next action"])
            lines = [f"# {headings[0]}", f"Organization: {organization or '[add]'}"]
            lines.extend(f"## {heading}\n{details if heading == 'Status and next action' and details else '[add]'}" for heading in headings[1:])
            return "\n\n".join(lines)

        @tool
        def search_internal_policies(query: str, documents: list[dict]) -> str:
            """Search supplied policy records and return matching text with supplied citations."""
            needle = query.strip().lower()
            matches = [doc for doc in documents or [] if needle in f"{doc.get('title', '')} {doc.get('text', '')}".lower()]
            if not matches:
                return "No supplied internal policy matched. Do not infer an organizational rule."
            return "\n".join(f"{doc.get('title', 'Untitled')} — {doc.get('text', '')} [Source: {doc.get('source', 'not supplied')}]" for doc in matches[:5])

        return [generate_nonprofit_template, search_internal_policies]

    if profile_id == "mutual-aid":
        @tool
        def sanitize_mutual_aid_intake(text: str) -> str:
            """Redact common contact, precise-location, and sensitive-status details."""
            redacted = re.sub(r"[\w.+-]+@[\w-]+\.[\w.-]+", "[redacted contact]", text or "")
            redacted = re.sub(r"(?:\+?\d[\d\s().-]{7,}\d)", "[redacted contact]", redacted)
            return re.sub(r"\b(?:address|location|diagnosis|medical status|immigration status|shelter)\s*:\s*[^;\n]+", lambda match: f"{match.group(0).split(':')[0]}: [redacted]", redacted, flags=re.IGNORECASE)

        @tool
        def pair_mutual_aid_needs(needs: list[dict], offers: list[dict]) -> str:
            """Pair anonymous needs and offers only after explicit opt-in from both sides."""
            matches = []
            for need in needs or []:
                if need.get("optIn") is not True:
                    continue
                tags = {str(tag).lower() for tag in need.get("tags", [])}
                offer = next((item for item in offers or [] if item.get("optIn") is True and tags.intersection(str(tag).lower() for tag in item.get("tags", []))), None)
                if offer:
                    matches.append({"needId": need.get("id", "anonymous-need"), "offerId": offer.get("id", "anonymous-offer"), "status": "pending-mutual-confirmation"})
            return json.dumps({"matches": matches, "note": "No contact details were shared. Both participants must confirm before connection."})

        return [sanitize_mutual_aid_intake, pair_mutual_aid_needs]

    if profile_id == "civic-knowledge":
        @tool
        def verify_civic_sources(sources: list[dict]) -> str:
            """Verify supplied official HTTPS sources before grounding civic answers."""
            checked = []
            for source in sources or []:
                url = str(source.get("url", ""))
                official = bool(re.search(r"\.(gov|gc\.ca|edu|org)(/|$)", url, re.IGNORECASE))
                checked.append({"title": source.get("title", "Untitled"), "url": url, "verified": source.get("verified") is True and url.startswith("https://") and official})
            grounded = bool(checked) and all(source["verified"] for source in checked)
            return json.dumps({"grounded": grounded, "sources": checked, "instruction": "If grounded is false, state uncertainty and provide an official contact point."})

        return [verify_civic_sources]

    return []
